### PR TITLE
Make client know about other solo players, fixes #258

### LINKED ISSUE #258: Proper prediction when one player is in solo, other not
This currently doesn't work becuase the client only knows whether the player itself is in solo, not about other ones. Bugs with /showothers 1

### PR DESCRIPTION
**Pull request by edg-l**
When another player is in solo and you are not, you could notice a weird collision and also the hook moving a little towards the solo player. This as been fixed.

It also applies m_ClShowOthersAlpha to other players whenever you are on solo or others are.

It uses the protocol extension thingy.

fixes #258

### CODE REVIEW

**Comment by Learath2** [comment]
I think this doesn't belong in an extended protocol message but in an extended snapshot element.

Also #392 comes to mind, maybe we should take another look at that.

**Comment by Ryozuki** [comment]
> I think this doesn't belong in an extended protocol message but in an extended snapshot element.
> 
> Also #392 comes to mind, maybe we should take another look at that.

git god heinrich told me on irc to use the extended protocol.

**Comment by Ryozuki** [comment]
For some reason prediction when antiping is on isnt fixed at all, but works perfect when its off. Idk if worth fixing with the massive rework of antiping.

https://cdn.discordapp.com/attachments/314776480880132097/566276262742720512/2019-04-12_16-58-13.mp4

**Comment by Learath2** [comment]
@Ryozuki He told you to use a netobj aswell :P
```
[9:22 PM] heinrich5991: hmm
[9:22 PM] heinrich5991: better to add a new netobj with the protocol extensions
[9:22 PM] heinrich5991: demos need that thing available in every tick

**Comment by heinrich5991** [comment]
Perhaps add a `CNetObj_DDRaceCharacter` with flags variable where this flag is the lowest bit.

**Comment by Learath2** [comment]
@heinrich5991 should we maybe think about a more comprehensive DDNetCharacter (#392)?

I don't really like the idea of piling these on without planning it out.

**Comment by heinrich5991** [comment]
Yes, I'd prefer an extensible `CNetObj_DDNetCharacter` as well. Make it have one flag for now (range min_int to max_int for future extensions) that currently specifies whether a tee is in solo or not.

**Comment by Ryozuki** [comment]
Oh yeah, if you do this i can probably fix #1595 too

**Comment by Learath2** [comment]
@Ryozuki want to do it or should I?

**Comment by Ryozuki** [comment]
> @Ryozuki want to do it or should I?

i dont want to create the ddnetcharacter, i want to use it to fix the things :)

File: datasrc/network.py [hunk 1]
```diff
@@ -56,7 +57,8 @@
 Flags = [
 	Flags("PLAYERFLAG", PlayerFlags),
 	Flags("GAMEFLAG", GameFlags),
-	Flags("GAMESTATEFLAG", GameStateFlags)
+	Flags("GAMESTATEFLAG", GameStateFlags),
+	Flags("CHARACTERFLAG", CharacterFlags)
```
Comment (heinrich5991) [inline datasrc/network.py]:
```suggestion
	Flags("CHARACTERFLAG", CharacterFlags),
```
Next diff doesn't need to touch two lines.

File: datasrc/network.py [hunk 2]
```diff
@@ -206,6 +208,10 @@
 		NetIntRange("m_AuthLevel", "AUTHED_NO", "AUTHED_ADMIN"),
 	]),
 
+	NetObjectEx("DDNetCharacter", "character@netobj.ddnet.tw", [
+		NetIntRange("m_Flags", "min_int", "max_int"),
```
Comment (heinrich5991) [inline datasrc/network.py]:
```suggestion
		NetIntAny("m_Flags"),
```

**Review by heinrich5991 — COMMENTED** [review]
LGTM, see nits.

File: datasrc/network.py (hunk 1 above)
Comment (fokkonaut) [reply datasrc/network.py]: why is that important? I think its more clear without that comma

Comment (def-) [reply datasrc/network.py]: easier to append a new flag, easier to resolve conflicts.

**Comment by heinrich5991** [comment]
Thanks!

bors r+

**Pull request merged**